support/releaseinfo/registryclient/client.go

```diff
@@ -223,36 +232,110 @@ func ExtractImageFilesToDir(ctx context.Context, imageRef string, pullSecret []b
+		return nil, reference.DockerImageReference{}, fmt.Errorf("failed to parse docker credentials: %w", err)
 	}
 	registryContext := registryclient.NewContext(rt, insecureRT).WithCredentials(credStore).
 		WithRequestModifiers(transport.NewHeaderRequestModifier(http.Header{http.CanonicalHeaderKey("User-Agent"): []string{rest.DefaultKubernetesUserAgent()}}))
 
 	ref, err := reference.Parse(imageRef)
 	if err != nil {
-		return nil, nil, fmt.Errorf("failed to parse image reference %q: %w", imageRef, err)
+		return nil, reference.DockerImageReference{}, fmt.Errorf("failed to parse image reference %q: %w", imageRef, err)
 	}
 	repo, err := registryContext.Repository(ctx, ref.DockerClientDefaults().RegistryURL(), ref.RepositoryName(), false)
 	if err != nil {
-		return nil, nil, fmt.Errorf("failed to create repository client for %s: %w", ref.DockerClientDefaults().RegistryURL(), err)
+		return nil, reference.DockerImageReference{}, fmt.Errorf("failed to create repository client for %s: %w", ref.DockerClientDefaults().RegistryURL(), err)
 	}
-	firstManifest, location, err := manifest.FirstManifest(ctx, ref, repo)
+	return repo, ref, nil
+}
+
+// Gets the manifest from a docker image
+func GetManifest(ctx context.Context, imageRef string, pullSecret []byte) (distribution.Manifest, error) {
+	repo, ref, err := GetRepoSetup(ctx, imageRef, pullSecret)
+
+	var srcDigest digest.Digest
+	if len(ref.ID) > 0 {
+		srcDigest = digest.Digest(ref.ID)
+	} else if len(ref.Tag) > 0 {
+		desc, err := repo.Tags(ctx).Get(ctx, ref.Tag)
+		if err != nil {
+			return nil, err
+		}
+		srcDigest = desc.Digest
+	} else {
+		return nil, err
+	}
+	manifests, err := repo.Manifests(ctx)
 	if err != nil {
-		return nil, nil, fmt.Errorf("failed to obtain root manifest for %s: %w", imageRef, err)
+		return nil, err
 	}
-	_, layers, err := manifest.ManifestToImageConfig(ctx, firstManifest, repo.Blobs(ctx), location)
+	digestsManifest, err := manifests.Get(ctx, srcDigest, manifest.PreferManifestList)
 	if err != nil {
-		return nil, nil, fmt.Errorf("failed to obtain image layers for %s: %w", imageRef, err)
+		return nil, err
+	}
+
+	return digestsManifest, nil
+}
+
+// Determines whether an image is a manifest listed image and contains manifests the following processor architectures: amd64, arm64, s390x, ppc64le
+func IsMultiArchManifestList(ctx context.Context, imageRef string, pullSecret []byte) (bool, error) {
+	srcManifest, err := GetManifest(ctx, imageRef, pullSecret)
+	if err != nil {
+		return false, fmt.Errorf("failed to retrieve manifest %s: %w", imageRef, err)
+	}
+
+	mediaType, payload, err := srcManifest.Payload()
+	if err != nil {
+		return false, fmt.Errorf("failed to get payload %s: %w", imageRef, err)
+	}
+
+	// This is the media type for manifest listed (aka fat manifest) images per the docker documentation - https://docs.docker.com/registry/spec/manifest-v2-2/
+	if mediaType == "application/vnd.docker.distribution.manifest.list.v2+json" {
```

Comment:
I'm not sure this check is reliable https://github.com/opencontainers/image-spec/blob/3a7f492d3f1bcada656a7d8c08f3f9bbd05e7406/media-types.md#applicationvndociimageindexv1json

https://github.com/opencontainers/image-spec/blob/main/manifest.md

Reply:
@flavianmissi and @dmage - @Prashanth684 suggested I reach out to you two about a question I had regarding programmatically figuring out if an image is manifest listed  or not. I have the code above to look at the mediaType to see if it was `application/vnd.docker.distribution.manifest.list.v2+json` which seems in line with the docker documentation [here](https://docs.docker.com/registry/spec/manifest-v2-2/). I think I’ve seen in oc that they just look at the length of the manifests? Is that the more appropriate way or is looking at that mediaType okay too?

My only nitpick I have about looking at the length is I have to deserialize the payload before I can actually see that. It would be nice if there was something said it was or was not a manifest list upfront. That mediaType seemed to be the closest thing to it I could find.

Reply:
media type is exactly what you need; there are two media types that represent manifest lists:

* application/vnd.docker.distribution.manifest.list.v2+json
* application/vnd.oci.image.index.v1+json

Reply:
Ok, thank you!